### ISSUE TITLE
On changing optimizers in the middle of training

### ISSUE THREAD

**Issue by songjiahao-wq** [ultralytics/yolov5#11291]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

How to implement adam in the first 100 rounds and SGD in the last 100 rounds

**Comment by glenn-jocher** [comment]
@songjiahao-wq to switch from one optimizer to another optimizer during training, you can utilize the `torch.optim.lr_scheduler` module provided by PyTorch. You can define two optimizers and in the `step` function, change the optimizer once the desired number of epochs is reached. Here's an example:

```python
import torch.optim.lr_scheduler as lr_scheduler

optimizer1 = torch.optim.Adam(model.parameters(), lr=0.001)
optimizer2 = torch.optim.SGD(model.parameters(), lr=0.01)

scheduler = lr_scheduler.MultiStepLR(optimizer1,
                                      milestones=[100],
                                      gamma=0.1)

for epoch in range(200):
    # train the model for one epoch
    if epoch == 100:
        optimizer1 = optimizer2
        scheduler = lr_scheduler.MultiStepLR(optimizer1,
                                              milestones=[150],
                                              gamma=0.1)
    scheduler.step()
```
In the example above, we define two optimizers `optimizer1` and `optimizer2`, where the first optimizer is Adam and the second is SGD. We then initialize the `MultiStepLR` scheduler with `optimizer1` and set a milestone to switch the optimizer from `optimizer1` to `optimizer2` after 100 epochs. Once the milestone is reached in the `step` function, we change the optimizer and re-initialize the scheduler with the new optimizer and new milestones to switch back to `optimizer1` after another 50 epochs.

I hope this helps!

**Comment by wengti** [comment]
Hi there, I am interested in implementing this change as well. Could you be more specific where should I insert these lines in my train.py?

**Comment by glenn-jocher** [comment]
@wengti sure, you can insert the lines for changing the optimizer in your `train.py` script at the appropriate location. Typically, you would place them just after the `optimizer` is initialized and before the training loop.

Here's an example of where you can add the code:

```python
optimizer1 = torch.optim.Adam(model.parameters(), lr=0.001)
optimizer2 = torch.optim.SGD(model.parameters(), lr=0.01)

scheduler = lr_scheduler.MultiStepLR(optimizer1,
                                      milestones=[100],
                                      gamma=0.1)

for epoch in range(200):
    # train the model for one epoch
    
    if epoch == 100:
        optimizer = optimizer2
        scheduler = lr_scheduler.MultiStepLR(optimizer,
                                              milestones=[150],
                                              gamma=0.1)
    scheduler.step()
```

The code above initializes `optimizer1` and `optimizer2` with the desired optimizers and learning rates. Then, inside the training loop, it checks if the desired epoch is reached (`epoch == 100`), and if so, it switches the `optimizer` to `optimizer2`. Adjust the epoch values and any other parameters according to your specific needs.

Let me know if you have any further questions!

**Comment by MSKBOX** [comment]
@glenn-jocher Hi there,I have two main issues regarding the current training process:

1.Training for just one epoch at a time results in the model outputting the network structure and parameters repeatedly, followed by a validation run on the validation set after each epoch. This leads to a significant waste of time, especially since I'm not interested in repeated validation at every single epoch.

2.With each new epoch, a new training folder is created, and new weight files are saved. This behavior is inefficient and not what I need. Ideally, I want the training to be continuous, without frequent interruptions for validation or unnecessary file saving.

I wonder if configuring this directly in the train.py script is sufficient, or if there are other underlying settings that I need to adjust to achieve seamless, uninterrupted training?Should I change the trainer.py?"there is my train.py 
`
from ultralytics import RTDETR
import torch
import torch.distributed as dist
import torch.optim as optim
import torch.optim.lr_scheduler as lr_scheduler
def main():

    yaml_file_path = '/root/detr/ultralytics/cfg/datasets/coco.yaml'

    model = RTDETR("/root/detr/ultralytics/cfg/models/a10test-r50.yaml")

    optimizer1 = optim.AdamW(model.model.parameters(), lr=0.0001, weight_decay=1e-4)
    optimizer2 = optim.SGD(model.model.parameters(), lr=0.01, momentum=0.9, weight_decay=1e-4)

    scheduler = lr_scheduler.MultiStepLR(optimizer1, milestones=[50], gamma=0.1)

    epochs = 100
    optimizer = optimizer1
    for epoch in range(epochs):

        if epoch == 2:
            optimizer = optimizer2
            scheduler = lr_scheduler.MultiStepLR(ptimizer, milestones=[80], gamma=0.1)
            print(f"\n切换优化器到 SGD，在 epoch {epoch} 之后\n")
        results = model.train(data=yaml_file_path,split='test',cls=0.5,box=7.5,dfl=1.5,overlap_mask=False,close_mosaic=0,mask_ratio=0,iou=0.5,epochs=1,cos_lr=True,weight_decay=0.0001,momentum=0.937,imgsz=640, batch=10,project='runs/train',hsv_h=0.015,hsv_s=0.7,hsv_v=0.4,degrees=0,translate=0,scale=0.5,fliplr=0.3,flipud=0.0,mosaic=0.1,mixup=0.0,copy_paste=0.0,name='exp')

        scheduler.step()   
if __name__ == "__main__":
    main()`
